Nextra 4 Blogpost

dimaMachina opened the pull request.

saihaj commented on website/pages/blog/nextra-4/index.mdx:
```diff
@@ -0,0 +1,1228 @@
+
+## App Router Support
+
+[Pages Router](https://nextjs.org/docs/pages) support has been removed, Nextra 4 is fully compatible
+with Next.js [App Router](https://nextjs.org/docs/app).
```
```suggestion
Nextra 4 exclusively uses Next.js [App Router](https://nextjs.org/docs/app). Support for the [Pages Router](https://nextjs.org/docs/pages) has been discontinued.
```

saihaj commented on website/pages/blog/nextra-4/index.mdx:
```diff
@@ -0,0 +1,1228 @@
+
+#### `content` Directory Convention
+
+The first mode is the easiest way to migrate your existing Nextra website from Pages router. To do
+follow these steps:
```
```suggestion
Migrate your Pages Router site with minimal changes using this mode. Here's what to do:
```

saihaj commented on website/pages/blog/nextra-4/index.mdx:
```diff
@@ -0,0 +1,1228 @@
+<Callout type='info'>
+**Note**
+
+In Web you can found article such as
+[Refreshing the Next.js App Router When Your Markdown Content Changes](https://steveruiz.me/posts/nextjs-refresh-content)
+where you need to install unnecessary dependency like `concurrently` and `ws`, mentioned article is
+adapted on
+[Dan Abramov workaround with `<AutoRefresh>` component and dev web socket server](https://github.com/hashicorp/next-remote-watch/issues/42#issuecomment-1794052655).
+Nextra's `content` mode provide ultimate DX:
+
+- you don't need to install unnecessary dependencies
+- you don't need to restart your server on changes in `content` directory
+- hot reloading works out of box
+- you can use `import` statements in MDX files and
+  [static images](https://nextra.site/docs/guide/image#static-image) works as well
+
+See Nextra's
+[example of docs website](https://github.com/shuding/nextra/tree/d5cb1bc0e84494cbcc8decbae56fb091ce9999f0/examples/docs)
+and
+[i18n website example](https://github.com/shuding/nextra/tree/d5cb1bc0e84494cbcc8decbae56fb091ce9999f0/examples/swr-site)
+as an examples of this mode.
```
```suggestion
Many existing solutions
[Refreshing the Next.js App Router When Your Markdown Content Changes](https://steveruiz.me/posts/nextjs-refresh-content), rely on extra dependencies like `concurrently` and `ws`. These approaches include 
[Dan Abramov workaround with `<AutoRefresh>` component and dev web socket server](https://github.com/hashicorp/next-remote-watch/issues/42#issuecomment-1794052655).

Nextra's content mode delivers a streamlined solution right out of the box: no extra dependencies, instant hot reloading, and full MDX support including working import statements and static images!

Checkout Nextra's [docs website](https://github.com/shuding/nextra/tree/d5cb1bc0e84494cbcc8decbae56fb091ce9999f0/examples/docs) and [i18n website example](https://github.com/shuding/nextra/tree/d5cb1bc0e84494cbcc8decbae56fb091ce9999f0/examples/swr-site).
```

saihaj commented on website/pages/blog/nextra-4/index.mdx:
```diff
@@ -0,0 +1,1228 @@
+
+#### `page` File Convention
+
+The second mode follows
+[Next.js App Router convention](https://nextjs.org/docs/getting-started/project-structure#app-routing-conventions)
+but with enhanced page extensions as `page.{md,mdx}` and don't requires additional steps to setup.
```
```suggestion
The second mode follows
[Next.js App Router convention](https://nextjs.org/docs/getting-started/project-structure#app-routing-conventions) with ✨ enhanced page extensions `page.{md,mdx}`.
```

saihaj commented on website/pages/blog/nextra-4/index.mdx:
```diff
@@ -0,0 +1,1228 @@
+
+## Turbopack Support
+
+After [more than 2 years of requesting this feature](https://github.com/shuding/nextra/issues/1054),
+Nextra 4 supports [Turbopack](https://nextjs.org/docs/architecture/turbopack), incremental bundler
+written in Rust. To benefit it, simple append the `--turbopack` flag to your development command:
```
```suggestion
After being one of our [most requested features](https://github.com/shuding/nextra/issues/1054) for over 2 years, Nextra 4 finally supports Turbopack, the Rust-based incremental bundler. Enable it by adding `--turbopack` to your dev command:
```

saihaj commented on website/pages/blog/nextra-4/index.mdx:
````diff
@@ -0,0 +1,1228 @@
+}
+```
+
+<Callout type="info">
+**Note**
+
+Without `--turbopack` flag Next.js under the hood uses
+[Webpack](https://github.com/webpack/webpack), written in JavaScript.
+
+</Callout>
````
I don't think it is really necessary. I have seen turbopack have some weird issues so would say webpack way is still the most stable option.   
```suggestion
```

dimaMachina replied:
Why? Just tip for users that without turbopack - JavaScript powered Webpack is used 😅

saihaj commented on website/pages/blog/nextra-4/index.mdx:
```diff
@@ -0,0 +1,1228 @@
+
+</Callout>
+
+## Drop Support of `theme.config` Files
```
```suggestion
## Discontinuing `theme.config` Support
```

dimaMachina commented on website/pages/blog/nextra-4/index.mdx:
```diff
@@ -0,0 +1,1228 @@
+1. For MDX pages while using `nextra-theme-docs` and `nextra-theme-blog` **you don't need to add**
+   `data-pagefind-body` attribute.
+
+</Callout>
```
changed

dimaMachina commented on website/pages/blog/nextra-4/index.mdx:
```diff
@@ -0,0 +1,1228 @@
+
+---
+
+[^1]: `import { Banner } from 'nextra/components'{:js}`
+
+[^2]: `import { Head } from 'nextra/components'{:js}`
+
+[^3]: `import { Navbar } from 'nextra-theme-docs'{:js}`
+
+[^4]: `import { Layout } from 'nextra-theme-docs'{:js}`
+
+[^5]: `import { LastUpdated } from 'nextra-theme-docs'{:js}`
+
+[^6]: `import { Footer } from 'nextra-theme-docs'{:js}`
+
+[^7]: `import { NotFoundPage } from 'nextra-theme-docs'{:js}`
+
+[^8]: `import { Search } from 'nextra/components'{:js}`
+
+[^9]: `import { Comments } from 'nextra-theme-blog'{:js}`
```
ty, it because table of migration was closed by default in `<details>` element, fixed in https://github.com/the-guild-org/website/pull/1803/commits/5471c6a511496ee01e263e1611e0a4b0e69359e0

saihaj commented on website/pages/blog/nextra-4/index.mdx:
````diff
@@ -0,0 +1,1315 @@
+}
+```
+
+## Conclusion. Why Prefer **Nextra 4** over Other MDX Libraries?
````
saying MDX libraries doesn't do justice since Nextra is more than MDX support :)

```suggestion
## Conclusion. Why Prefer **Nextra 4** over others?
```

saihaj commented on website/pages/blog/nextra-4/index.mdx:
```diff
@@ -0,0 +1,1315 @@
+
+## Remote Docs
+
+Configuration for remote docs was changed, please refers to
+[official example](https://github.com/shuding/nextra/blob/aca79fa5e224e861c253d0cae5bed91268804a19/examples/swr-site/app/%5Blang%5D/remote/graphql-eslint/%5B%5B...slug%5D%5D/page.tsx),
+also you will need to manually modify `pageMap` list in `layout` file, to properly display sidebar
+navigation links. Example of modified `layout` file is
+[here](https://github.com/shuding/nextra/blob/aca79fa5e224e861c253d0cae5bed91268804a19/examples/swr-site/app/%5Blang%5D/layout.tsx#L52-L56).
```
```suggestion
Remote docs configuration has changed, please refers to
[official example](https://github.com/shuding/nextra/blob/aca79fa5e224e861c253d0cae5bed91268804a19/examples/swr-site/app/%5Blang%5D/remote/graphql-eslint/%5B%5B...slug%5D%5D/page.tsx),
Additionally you will need to modify `pageMap` list in `layout` file, to properly display sidebar
navigation links. Example of modified `layout` file is
[here](https://github.com/shuding/nextra/blob/aca79fa5e224e861c253d0cae5bed91268804a19/examples/swr-site/app/%5Blang%5D/layout.tsx#L52-L56).
```

saihaj commented on website/pages/blog/nextra-4/index.mdx:
```diff
@@ -0,0 +1,1315 @@
+  </FileTree.Folder>
+</FileTree>
+
+#### List Subpages
+
+It's now possible to automatically list all subpages of a given route as cards.
```
Part of the code I think we should also so how it would look like in preview. You could add a screenshot alternatively just link to an example of some deployment where we can see this feature in action :)

dimaMachina replied:
fixed in https://github.com/the-guild-org/website/pull/1803/commits/bdb9ff0d06bf7302190ce890f3203e8b206cea76

saihaj commented on website/pages/blog/nextra-4/index.mdx:
```diff
@@ -0,0 +1,1315 @@
+The Nextra 4 documentation is still a work in progress, and we need your help to improve it!<br />
+Check out the recent updates in this blog post, and feel free to submit your contribution PRs to improve Nextra docs.
+
+Spread the word about Nextra on [X](https://x.com) using the `#Nextra` hashtag. You can also follow
+me on [X](https://x.com/dimaMachina_) and [GitHub](https://github.com/dimaMachina) to stay updated!
```
```suggestion
Spread the word about Nextra on [𝕏](https://x.com) using the `#Nextra` hashtag. You can also follow
me on [𝕏](https://x.com/dimaMachina_) and [GitHub](https://github.com/dimaMachina) to stay updated!
```

87xie commented:
In v3, users might use `useRouter` from `nextra/hooks`. Should we remind them about it here?

87xie commented on website/pages/blog/nextra-4/index.mdx:
```diff
@@ -0,0 +1,1315 @@
+[Next.js App Router convention](https://nextjs.org/docs/getting-started/project-structure#app-routing-conventions)
+with ✨ enhanced page extensions `page.{md,mdx}`.
+
+Above [`content` directory convention structure](#add-mdxpathpagejsx-file), with this mode will
+looks like:
```
It looks like there's a typo in the url of the link

dimaMachina replied:
nice catch, it because I removed Markdown heading (to avoid adding steps in TOC) and replaced by h5, fixed with 
<img width="603" alt="image" src="https://github.com/user-attachments/assets/4d9c2ba6-ab8f-481c-89c7-964b35545961" />

dimaMachina replied:
fixed in https://github.com/the-guild-org/website/pull/1803/commits/bdb9ff0d06bf7302190ce890f3203e8b206cea76

87xie commented on website/pages/blog/nextra-4/index.mdx:
```diff
@@ -0,0 +1,1315 @@
+
+### Code Block Icons Changes
+
+#### Overriding Icons
+
+You can now override or add additional icons for code blocks. Here is an example how to do it for
+`nextra-theme-docs`, but it's the same approach for custom themes. Import `withIcons` HOC function
+from `nextra/components`, Wrap the `Pre` component with `withIcons`, passing in your custom icon for
+the desired language.
```
```suggestion
#### Customizing Icons

You can now customize icons for code blocks. Here's an example for `nextra-theme-docs`, but the same approach applies to custom themes:

1. Import `withIcons` HOC function from `nextra/components`.
1. Wrap the `Pre` component with `withIcons`, passing your custom icon for the desired language.
```

I think the word "customize" is more appropriate than "override," and listing the steps might make it easier to read.

dimaMachina replied:
nice, looks better!

87xie commented on website/pages/blog/nextra-4/index.mdx:
````diff
@@ -0,0 +1,1315 @@
+}
+```
+
+#### JSX Icons
+
+The `jsx` and `tsx` languages now display the React icon for better visual identification.
+
+#### Diff Icons
+
+For code blocks with the diff language, if a
+[`filename` attribute](https://nextra.site/docs/guide/syntax-highlighting#filenames-and-titles) is
+provided, the icon will automatically correspond to the file extension of the specified filename.
````
```suggestion
#### Icons Updates

- **JSX icons**: The `jsx` and `tsx` languages now display the React icon for better visual identification.
- **Diff icons**: For code blocks with the `diff` language, if a
[`filename` attribute](https://nextra.site/docs/guide/syntax-highlighting#filenames-and-titles) is
provided, the icon will automatically correspond to the file extension of the specified filename.
```

I think categorizing under "Icons Updates" would make the TOC more intuitive.

dimaMachina replied:
ty

dimaMachina commented:
> Some of my thoughts:
> 
> * In v3, users might use `useRouter` from `nextra/hooks`, but it has been deprecated in v4. Do you think we should remind them about it here?
> * I’ve seen several questions about [dynamic head tags](https://nextra-v2-7hslbun8z-shud.vercel.app/docs/docs-theme/theme-configuration#dynamic-tags-based-on-page), particularly regarding the title template. Since the `head` option in `theme.config` has been removed in v4, it might be helpful to include an example in the migration guide.
> * I prefer to reduce the heading levels to make the TOC content easier to read. I'm not sure if this change would be helpful, so I created a branch and made some adjustments to the "App Router Support" [here](https://github.com/87xie/website/tree/adjustments-for-the-app-router-section).
> * I think a new "Trade-offs" heading could be added to the "App Router Support" to help users make better decisions. For example:
>   
>   * The `catch-all` route might require longer compilation times depending on the number of MDX files.
>   * The page file convention works well with colocation, allowing you to keep all the assets for an article in one place.

Thanks @87xie I addressed all of your suggestions, and did a cherry pick of your commit
